posts/MarkoB/2017-02-06-validator-service-loader.adoc

```diff
@@ -0,0 +1,290 @@
+
+* `initialize` - initializes validator based on annotation parameters.
+* `isValid` - performs actual validation
+
+A possible implementation can be next:
+
+```java
+public class DurationMinValidator implements ConstraintValidator<DurationMin, Duration> {
+
+	private Duration duration;
+
+	@Override
+	public void initialize(DurationMin constraintAnnotation) {
+		this.duration = Duration.of( constraintAnnotation.value(), constraintAnnotation.units() );
+	}
+
+	@Override
+	public boolean isValid(Duration value, ConstraintValidatorContext context) {
+		// null values are valid
+		if ( value == null ) {
+			return true;
+		}
+		return duration.compareTo( value ) < 1;
+	}
+}
+```
+
+As was mentioned earlier - you want to hide this implementation from API users by putting it somewhere in internal
+packages, in our case it will be - `org.hibernate.validator.internal.constraintvalidators.hv.time`.
+
+As we are creating a new constraint annotation we should also provide a default message for it as well. This can be
+done by placing `ContributorValidationMessages.properties` property file on a classpath. This property file should
+contain a key/message pair, where key is the one used in annotation declaration (in our case it's
+`org.hibernate.validator.constraints.time.DurationMin.message`). And message is the one you would like to show
+when validation fails. A possible content of this property file in our case can be:
+
+	org.hibernate.validator.constraints.time.DurationMin.message = must be greater than or equal to {value} {units}
+
+If you leave everything else as is your constraint annotation will live its own life without knowing about a presence
+of validator. Hibernate Validator will not know about your `DurationMinValidator` as well. So to make sure that
+Hibernate Validator find your `DurationMinValidator` you need to create `META-INF/services/javax.validation.ConstraintValidator`
+file and put a fully qualified name of validator you want to be exposed to Hibernate Validator:
+
+	#providing a list of validators for new @DurationMin annotation
+    org.hibernate.validator.internal.constraintvalidators.hv.time.DurationMinValidator
+
+To make sure that it all worked out let's run a test:
+
+```java
+Some cool test case will appear here later ;)
+```
+
+All code can for this example can be found at http://some-valid-github-link-to-hibernate-demo-repo-will-be-here[GitHub]
+
+== Messages
+
+This mechanism can also be used in structuring your custom constraint messages. Instead of having them all
+together you can group them by modules and they will be picked up and used by Hibernate Validator. To achive this
+just make sure that each module that you'd like to have custom messages have `ContributorValidationMessages.properties`
+property file which contains key/message pairs for those constraints that are needed in that particular module.
```

Comment:
This sentence is too long and hard to follow. To be honest, I'm not even sure this additional paragraph is needed. You already covered a lot of things.